partners_api/uber_api.cpp

```diff
@@ -0,0 +1,221 @@
+
+#include "private.h"
+
+using namespace platform;
+
+namespace
+{
+uint32_t const kHttpMinWait = 10;
+
+string RunSimpleHttpRequest(string const & url)
+{
+  HttpClient request(url);
+  if (request.RunHttpRequest() && !request.WasRedirected() && request.ErrorCode() == 200)
+  {
+    return request.ServerResponse();
+  }
+  return {};
+}
+
+/// Feature should refers to a shared state.
+template <typename T>
+bool WaitForFeature(future<T> const & f, uint32_t waitMillisec, atomic<bool> const & runFlag)
+{
+  future_status status = future_status::deferred;
+  while (runFlag && status != future_status::ready)
+  {
+    status = f.wait_for(milliseconds(waitMillisec));
+  }
+
+  return runFlag;
+}
+
+bool CheckUberAnswer(json_t const * answer)
+{
+  // Uber products are not available at this point.
+  if (!json_is_array(answer))
+    return false;
+
+  if (json_array_size(answer) <= 0)
+    return false;
+
+  return true;
+}
+
+void FillProducts(json_t const * time, json_t const * price, vector<uber::Product> & products)
+{
+  // Fill data from time.
+  auto const timeSize = json_array_size(time);
+  for (size_t i = 0; i < timeSize; ++i)
+  {
+    uber::Product product;
+    json_int_t estimatedTime = 0;
+    auto const item = json_array_get(time, i);
+    my::FromJSONObject(item, "display_name", product.m_name);
+    my::FromJSONObject(item, "estimate", estimatedTime);
+    product.m_time = strings::to_string(estimatedTime);
+    products.push_back(product);
+  }
+
+  // Fill data from price.
+  auto const priceSize = json_array_size(price);
+  for (size_t i = 0; i < priceSize; ++i)
+  {
+    string name;
+    auto const item = json_array_get(price, i);
+
+    my::FromJSONObject(item, "display_name", name);
+    auto const it = find_if(products.begin(), products.end(), [&name](uber::Product const & product){
+      return product.m_name == name;
+    });
+    if (it == products.end())
+      continue;
+
+    my::FromJSONObject(item, "product_id", it->m_productId);
+    my::FromJSONObject(item, "estimate", it->m_price);
+
+    // The field currency_code can contain null in case when price equal to Metered.
+    auto const currency = json_object_get(item, "currency_code");
+    if (currency != nullptr && !json_is_null(currency))
+      it->m_currency = json_string_value(currency);
+  }
+
+  products.erase(remove_if(products.begin(), products.end(), [](uber::Product const & p){
+                   return p.m_name.empty() || p.m_productId.empty() || p.m_time.empty() ||
+                          p.m_price.empty();
+                 }), products.end());
+}
+
+void GetAvailableProductsAsync(ms::LatLon const from, ms::LatLon const to,
+                               size_t const requestId, atomic<bool> const & runFlag,
+                               uber::ProductsCallback const fn)
+{
+  LOG(LINFO, ("uber::Api::GetAvailableProductsAsync begin"));
+  auto time = async(launch::async, uber::RawApi::GetEstimatedTime, ref(from));
```

Comment:
As I proposed, you do not need all these machinery with aux thread. Just start two async's and pass callbacks to them. On simple http requests completion, callbacks must be posted on the gui thread.